ray_uint8_patch.sh

```diff
@@ -0,0 +1,12 @@
+# Find Python folder name so that this patch can run correctly on different versions of Python.
+python_folder_name=$(ls venv/lib)
+
+# Apply patches
+sed -i '119s/tf.float32/tf.uint8/' venv/lib/"$python_folder_name"/site-packages/ray/rllib/policy/dynamic_tf_policy.py # Hardcoded observation space to uint8.
+sed -i '76s/np.float32/np.uint8/' venv/lib/"$python_folder_name"/site-packages/ray/rllib/models/preprocessors.py # Same as above.
+sed -i '231s/np.zeros(self.shape)/np.zeros(self.shape, dtype=self.observation_space.dtype)/' venv/lib/"$python_folder_name"/site-packages/ray/rllib/models/preprocessors.py # Change observation shape to what we actually provide
+sed -i '214s/tf.int64/action_space.dtype/' venv/lib/"$python_folder_name"/site-packages/ray/rllib/models/catalog.py # Change action shape to what we actually provide
+sed -i '56s/tf.math.argmax(self.inputs, axis=1)/tf.math.argmax(self.inputs, axis=1, output_type=tf.int32)/' venv/lib/"$python_folder_name"/site-packages/ray/rllib/models/tf/tf_action_dist.py # Actions should not sample at int64, int32 is the lowest that multinomial takes
+sed -i '84s/tf.multinomial(self.inputs, 1)/tf.multinomial(self.inputs, 1, output_dtype=tf.int32)/' venv/lib/"$python_folder_name"/site-packages/ray/rllib/models/tf/tf_action_dist.py # Same as above
+sed -i '656i\        actions = np.array(actions, dtype=policy.action_space.dtype)' venv/lib/"$python_folder_name"/site-packages/ray/rllib/evaluation/sampler.py # Insert action to uint8 conversion to save even more memory
+sed -i '164i\        return self.sess.run(self.variables)' venv/lib/"$python_folder_name"/site-packages/ray/experimental/tf_utils.py # Make initialization faster, as in https://github.com/ray-project/ray/pull/8491
```

Comment:
can you clarify what this is for?

Reply:
Added clarification in the file.

Reply:
Oh I see. Does the code still run without this or is this just for a speedup?

Reply:
It won't run without it, and it is for a speedup. The last line reduces initialization time by up to 60%, and training time by <20% as well (although I haven't tested the last part extensively, might just be a fluke).

The rest of the lines make observations 4 times smaller, but ray normally doesn't support this and casts everything to float32. Note that this replaces one hard-coding with another, which is an ugly hack - but fixing ray would require restructuring in many places, I feel like I'm out of my depth there.

Reply:
I think this is a great idea, I just worry that this hard-codes us to a particular ray version in a very difficult way (for example, if the ray version changes even slightly the patch file is a haul to fix). I don't have any good ideas yet but if you have any thoughts on a way to make this more modular let me know.

Reply:
I fully agree, it's an ugly hack. Any future-proof non-hacky solution would have to fix https://github.com/ray-project/ray/issues/7946. That's the only way to guarantee future compatibility without extra work on each patch.

The last line has already been patched in ray 0.8.6, by yours truly.